gap/semigroups/semifp.gi

```diff
@@ -636,3 +636,163 @@ function(M)
 
   return MagmaIsomorphismByFunctionsNC(M, MF, map, inv);
 end);
+
+InstallMethod(ParseRelations,
+"for a list of free generators and a string",
+[IsDenseList, IsString],
+function(gens, inputstring)
+    local newinputstring, g, chartoel, RemoveBrackets, ParseRelation, output;
+
+    for g in gens do
+      if not (Size(String(g)) = 1 and String(g)[1]
+         in "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ") then
+        ErrorNoReturn(Concatenation(
+        "expected a single english letter but found ", String(g)));
+      fi;
+    od;
+
+    newinputstring := Filtered(inputstring, x -> not x = ' ');
+    for g in gens do
+      newinputstring := ReplacedString(newinputstring,
+                        [String(g)[1], '^'], ['(', String(g)[1], ')', '^']);
+    od;
+
+    RemoveBrackets := function(word)
+        local i, product, lbracket, rbracket, nestcount, index, p, chartoel;
+        if word = "" then
+            ErrorNoReturn(Concatenation("expected a product of free",
+                          " generators but found an empty string"));
+        fi;
+
+        # if the number of left brackets is different from the number of right
+        # brackets they can't possibly pair up
+        if not Size(Filtered(word, x -> x = '(')) =
+               Size(Filtered(word, x -> x = ')')) then
+            ErrorNoReturn(Concatenation("expected the number of open brackets",
+                          " to match the number of closed brackets"));
+        fi;
+
+        # if the ^ is at the end of the string there is no exponent.
+        # if the ^ is at the start of the string there is no base.
+        if word[1] = '^' then
+            ErrorNoReturn(Concatenation("expected ^ to be preceded by a ) or",
+                          " a generator but found begining of string"));
+        elif word[Size(word)] = '^' then
+            ErrorNoReturn(Concatenation("expected ^ to be followed by a ",
+                          "positive integer but found end of string"));
+        fi;
+        # checks that all ^s have an exponent.
+        for index in [1 .. Size(word)] do
+            if word[index] = '^' then
+                if not word[index + 1] in "0123456789" then
+                  ErrorNoReturn(Concatenation("expected ^ to be followed by",
+                  " a positive integer but found ", [word[index + 1]]));
+                fi;
+                if word[index - 1] in "0123456789^(" then
+                  ErrorNoReturn(Concatenation(
+                  "expected ^ to be preceded by a ) or a generator",
+                  " but found ", [word[index - 1]]));
+                fi;
+            fi;
+        od;
+
+        # converts a character to the element it represents
+        chartoel := function(char)
```

Comment:
It's unlikely that this is ever going to be the bottleneck in anything, but this could be a lookup table instead.